akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1772,14 +1808,34 @@ trait OutHandler {
   @throws(classOf[Exception])
   def onPull(): Unit
 
+  // Hack to make sure that old `onDownstreamFinish` can be called without losing the cause in the default implementation
+  private[this] var _lastCancellationCause: Throwable = _
+
   /**
    * Called when the output port will no longer accept any new elements. After this callback no other callbacks will
    * be called for this port.
    */
   @throws(classOf[Exception])
+  // FIXME: add this after fixing our own usages
+  // @deprecatedOverriding("Override `def onDownstreamFinish(cause: Throwable)`, instead.", since = "2.6.0") // warns when overriding
```

Comment:
Added https://github.com/akka/akka/issues/27472 to follow up with this in an extra step. Until then some of our stages don't propagate cancellation causes.